### ISSUE TITLE
Stealing focus from TextBox with Validator

### ISSUE THREAD

**Issue by isushik94** [linebender/druid#1601]
Hi
I think it will be useful to add `request_focus` call to other controls (i.e a button) to steal it from an active text box to trigger validation and data updating.
The current behavior I seem to be user unfriendly as after entering some text the user is required to press Enter key instead of just clicking a button "create"/"next".

**Comment by cmyr** [comment]
I agree that this whole system needs some work. I'd like to get the IME stuff landed first, since that strongly influences the control model for textboxes, and then we can look into how textboxes should interact with other components.

**Comment by cmeissl** [comment]
I faced the same issue in a small tool i created.
After checking what is going on i came up with the following workaround.

```rust
struct FocusController {}

impl<D, W: Widget<D>> Controller<D, W> for FocusController {
    fn event(
        &mut self,
        child: &mut W,
        ctx: &mut druid::EventCtx,
        event: &druid::Event,
        data: &mut D,
        env: &Env,
    ) {
        /* This is a workaround for the following issue.
         * The text input with validation will validate only if
         * it looses focus. On clicking the button the focus
         * is still on the text input thus not setting the value
         * before invoking the click handler which will then
         * not use the correct values.
         * As a workaround we steal the focus of the text input when
         * the mouse moves above the button before the user clicks.
         * Stealing the focus in the click handler is too late.
         */
        if let Event::MouseMove(_) = event {
            ctx.request_focus();
        }
        child.event(ctx, event, data, env)
    }
}

```

The controller then needs to be added to the button with the controller extension method.

**Comment by cmyr** [comment]
I think this might point to a bigger issue, which is that clicking on controls should generally steal focus; currently we sort of only use focus for things that accept text-input, but that doesn't sound quite right

**Comment by jjl** [comment]
I've run into this one today as well, but in a different scenario - focusing a newly added  widget (loosely, a uml-like diagram editor). It's pretty horrible because the natural place to deal with it in the absence of being able to request focus on behalf of other controls `seems` during lifecycle, which can't request the focus...

**Comment by jjl** [comment]
Looks like there is a `register_for_focus` method on the lifecyclectx. i see it does something different from the `request_focus` of EventCtx, but i'm not clear on what or why it should have a different name

**Comment by cmyr** [comment]
`register_for_focus` is used if a widget wants to participate in the 'focus chain', that is if it wants to be eligible to gain focus when the user presses `tab` from another widget.

**Comment by cmyr** [comment]
For a newly added widget, the easiest thing to do is send that widget a `Command` and have it request focus in response. 🤷

**Comment by jjl** [comment]
Ah right. I had thought of that, but the reason I was inclined not to is because it seemed like needless indirection - the widget is right here in self, so the command routing process seems like (probably not much) added overhead.
